File: AlgoXpert/c++/stack_using_linked_list.cpp
#include <bits/stdc++.h> 
using namespace std; 

struct Node { 
	int data; 
	struct Node* link; 
}; 
struct Node* top; 
void push(int data) 
{ 
	struct Node* temp; 
	temp = new Node(); 
	if (!temp) { 
		cout << "\nHeap Overflow"; 
		exit(1); 
	} 
	temp->data = data; 
	temp->link = top; 
	top = temp; 
} 

int isEmpty() 
{ 
	return top == NULL; 
}  
int peek() 
{ 
	if (!isEmpty()) 
		return top->data; 
	else
		exit(1); 
} 
void pop() 
{ 
	struct Node* temp; 
	if (top == NULL) { 
		cout << "\nStack Underflow" << endl; 
		exit(1); 
	} 
	else { 
		temp = top; 
		top = top->link; 
		temp->link = NULL; 
		free(temp); 
	} 
} 

void display() 
{ 
	struct Node* temp; 
	if (top == NULL) { 
		cout << "\nStack Underflow"; 
		exit(1); 
	} 
	else { 
		temp = top; 
		while (temp != NULL) { 
			cout << temp->data << " "; 
			temp = temp->link; 
		} 
	} 
} 

int main() 
{ 
	push(110); 
	push(2); 
	push(78); 
	push(56); 
	display();  
	cout << "\nTop element is " << peek() <<"\n"; 
	pop(); 
	pop(); 
	display(); 
	cout << "\nTop element is " << peek() <<"\n"; 
	return 0; 
} 

File: AlgoXpert/dart/stack_using_linked_list.dart
class Node<T>{
  T data;
  Node next;
  
  Node(T data){
    this.data = data;
    this.next = null;
  }
}

class LinkedListStack<T>{
  Node<T> head;
  int size;
  
  LinkedListStack(){
    this.head = null;
    this.size = 0;
  }
  void push(T element){
       Node<T> newNode = new Node<T>(element);
       newNode.next = this.head;
       this.head = newNode;
       this.size++;
  }
  T pop(){
    T returnData = null;
    if(size == 0){
      print("The stack is empty!!!");
    }
    else{
      Node<T> destroy = this.head;
      this.head = this.head.next;
      returnData = destroy.data;
      this.size--;
    }
    return returnData;
  }
  
  bool isEmpty(){
    return this.size == 0;
  }
  
  int getSize(){
    return this.size;
  }
}

int main(){
  LinkedListStack<String> Stack = new LinkedListStack<String>();
  var returnData;
  print("Push 2 5 9 7 to the stack\n");
  Stack.push("2");
  Stack.push("5");
  Stack.push("9");
  Stack.push("7");
  print("Successful push!\n");
  returnData = Stack.pop();
  print("Pop a data: $returnData\n");
  returnData = Stack.pop();
  print("Pop a data: $returnData\n");
  returnData = Stack.pop();
  print("Pop a data: $returnData\n");
  return 0;
}